File: src/main/java/com/tristanbomb/powerofpaint/entity/EntityWaterBomb.java
package com.tristanbomb.powerofpaint.entity;

import java.util.ArrayList;
import java.util.List;

import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.projectile.EntityThrowable;
import net.minecraft.util.EnumParticleTypes;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.world.World;

public class EntityWaterBomb extends EntityThrowable
{
	int VRADIUS = 1;
	int HRADIUS = 3;
	
	int meta = 0;
	
    public EntityWaterBomb(World worldIn)
    {
        super(worldIn);
    }

    public EntityWaterBomb(World worldIn, EntityLivingBase throwerIn)
    {
        super(worldIn, throwerIn);
    }

    public EntityWaterBomb(World worldIn, double x, double y, double z)
    {
        super(worldIn, x, y, z);
    }

    /**
     * Called when this EntityThrowable hits a block or entity.
     */
	protected void onImpact(RayTraceResult result)
    {
		if (result.getBlockPos() == null) { return; }
		
        List<BlockPos> checkArray = new ArrayList<BlockPos>();
        
        for (int i = -VRADIUS; i <= VRADIUS; i++) { //Generate a list of block positions in a cylinder
			for (int j = -HRADIUS; j <= HRADIUS; j++) {
				for (int k = -HRADIUS; k <= HRADIUS; k++) {
					if (Math.pow(j,2) + Math.pow(k,2) < Math.pow(HRADIUS+0.5,2)) {
						int x = result.getBlockPos().getX() + j;
						int y = result.getBlockPos().getY() + i;
						int z = result.getBlockPos().getZ() + k;
						
						BlockPos bp = new BlockPos(x,y,z);
						checkArray.add(bp);
					}
				}
			}
		}
        
        for (int i = 0; i < checkArray.size(); i++) {
			BlockPos bp = checkArray.get(i);
			if (world.getBlockState(bp).getBlock().isReplaceable(world, bp)) {
				world.setBlockToAir(bp);
			}
		}
        
        for (int j = 0; j < 8; ++j)
        {
            this.world.spawnParticle(EnumParticleTypes.SNOWBALL, this.posX, this.posY, this.posZ, 0.0D, 0.0D, 0.0D, new int[0]);
        }

        if (!this.world.isRemote)
        {
            this.setDead();
        }
    }
}

File: src/main/kotlin/tristanbomb/powerofpaint/entity/EntityWaterBomb.kt
package tristanbomb.powerofpaint.entity

import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.projectile.EntityThrowable
import net.minecraft.util.EnumParticleTypes
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.RayTraceResult
import net.minecraft.world.World
import java.util.ArrayList
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.common.registry.EntityRegistry.registerModEntity
import tristanbomb.powerofpaint.PowerOfPaint
import kotlin.math.roundToInt


class EntityWaterBomb: EntityThrowable {
    val vRadius = 1
    val hRadius = 3
    constructor(worldIn: World) : super(worldIn) {}
    constructor(worldIn: World, throwerIn: EntityLivingBase) : super(worldIn, throwerIn) {}
    constructor(worldIn: World, x: Double, y: Double, z: Double) : super(worldIn, x, y, z) {}

    companion object {
        val resourceName = "water_bomb"
        val properName = "WaterBomb"
        @JvmStatic
        fun register(id: Int) {
            registerModEntity(ResourceLocation(PowerOfPaint.MOD_ID + ":" + resourceName), EntityWaterBomb::class.java, properName, id, PowerOfPaint, 64, 10, true)
        }
    }


    override fun onImpact(result: RayTraceResult) {
        val checkArray = ArrayList<BlockPos>()
        PowerOfPaint.info(result.toString())

        if (result.typeOfHit == RayTraceResult.Type.BLOCK) {
            for (i in -vRadius..vRadius) { //Generate a list of block positions in a cylinder
                for (j in -hRadius..hRadius) {
                    for (k in -hRadius..hRadius) {
                        if (Math.pow(j.toDouble(), 2.0) + Math.pow(k.toDouble(), 2.0) < Math.pow(hRadius + 0.5, 2.0)) {
                            val x = (result.blockPos.x) + j
                            val y = (result.blockPos.y) + i
                            val z = (result.blockPos.z) + k
                            checkArray.add(BlockPos(x, y, z))

                            this.world.spawnParticle(EnumParticleTypes.CLOUD, x + 0.5, y + 0.5, z + 0.5, 0.0, 0.0, 0.0, 10, 10, 10)
                        }
                    }
                }
            }

            for (i in checkArray.indices) {
                val bp = checkArray[i]
                if (world.getBlockState(bp).block.isReplaceable(world, bp) && !world.getBlockState(bp).material.isLiquid) {
                    world.setBlockToAir(bp)
                }
            }

            for (j in 0..10) {
                this.world.spawnParticle(EnumParticleTypes.EXPLOSION_NORMAL, this.posX, this.posY, this.posZ, 0.0, 0.0, 0.0, 500, 500, 500)
            }

            if (!this.world.isRemote) {
                this.setDead()
            }
        }
    }
}
